### DISCUSSION TITLE
Is the  "param1" of the call receipt no longer the function selector on the testnet?

### DISCUSSION THREAD

**Discussion by spacedragon** [FuelLabs/fuels-ts#2424] [Q&A] [answered]
<img width="506" alt="image" src="https://github.com/FuelLabs/fuels-ts/assets/631261/d053b6a3-4f34-4068-b913-ae83dfdddd75">

How can we find out what function is called from the receipts?

**Comment by maschad** [comment]
Could you share a code snippet including where you are calling this code from and also the the full receipt?

**Comment by spacedragon** [comment]
sway code
```
 abi Counter {
    #[storage(read, write)]
    fn increment(add: u32, two: u64, three: b256) -> u64;

    #[storage(read)]
    fn count() -> u64;

    #[storage(read, write)]
    fn complex(foo: Foo, option: Option<str[1]>) -> Result<Foo, ColorError>;
}
```

client code

```
try {
      const result = await contract.functions
      .complex({bar: 7, baz: true }, "r")
      .txParams({
        gasLimit: 100_000,
      })
      .call()

```
 

transaction on explorer
https://next-app.fuel.network/tx/0xc6c75b1bd6896a596123ad3447725b5e84861bf5568852c4426a24f176121755/simple

**Reply by maschad** [reply]
Thanks for sharing, unfortunately you are right, there were some changes in how the [contract calls are encoded](https://github.com/FuelLabs/sway/commit/c24d7319ab4b983a3abe59b6a6d99e09450a45ba) for testnet and so we need to make some modifications to our API endpoint to return values that will allow you to determine the function selector.

I've raised an issue on sway core here https://github.com/FuelLabs/sway/issues/6115

**Reply by maschad** [reply, marked as answer]
@spacedragon I must apologize as  this was overcomplicated.

I presume you are using `v1` encoding, thus the value `10480` returned by `param1` includes the base offset for the `scriptData` , and thus you do have access to the encoded function selector.  Therefore you should be able to determine which function is called.

Given `WORD_SIZE = 8` , `CONTRACT_ID_LEN = 32` and `ASSET_ID_LEN = 32` bytes, you can decode it as such:

```
const encodedFunctionSelectorOffset =  WORD_SIZE + ASSET_ID_LEN + CONTRACT_ID_LEN + WORD_SIZE + WORD_SIZE;
```

Therefore we can use the `scriptData` from the `TransactionResult` to get the `decodedFunctionSelector`

```ts
    // Submit the contract call
    const response = await contracts.functions.complex({bar: 7, baz: true }, "r")
      .txParams({
        gasLimit: 100_000,
      })
     .call()
    // Destruct the transaction result
    const { transactionResult } = response;
    // Get the script data from the transaction result
    const scriptData = arrayify(transactionResult.transaction.scriptData);
    // Decode the data as `String`, slicing by the selector offset
    const coder = new StdStringCoder();
    // Second parameter is the offset where it will slice the scriptData
    const [decodedFunctionSelector] = coder.decode(scriptData, encodedSelectorOffset);
   // decodedFunctionSelector === 'complex'

 ```

Do this solve your problem?

**Reply by spacedragon** [reply]
Thanks, will try.

**Reply by spacedragon** [reply]
Will fuel-ts sdk or API change this to make it easier?

**Reply by danielbate** [reply]
As of [V1 Encoding](https://docs.fuel.network/docs/specs/abi/argument-encoding/#version-1), we will not be able to change the value of `param1` as it is now a pointer to the encoded selector. 

What do you think would make this easier? As in helpers to decode the various parts of a contract call script?

We may be able to include the call data (`scriptData` in this context) in the contract call reciept, so you don't need to fetch the tx, would that help at all?

Also, the caveat here is that we can't always guarantee the encoding that is being used with a contract call. If a user has implemented their own encoding, we can't guarantee the value of the `param1`.

**Reply by spacedragon** [reply]
Yes, and also provide methods in fuel-ts so users do not have to handle this complexity.

**Reply by danielbate** [reply]
I've added #2530 and #2531 to handle these.
